TupleMap type

vtjnash opened the pull request:
The purpose of this change set is to separate the method caching logic used by generic functions into a independent utility. The motivation for this changes was #265 and (more recently, but related) #15555. Since both require the compiler to be able to invalidate various method caches, it seemed most practical to me to have this functionality available independent of generic function dispatch (so that dispatch is a consumer of this logic and information instead of the provider).

For reviewing, I recommend starting from the definition of the new TypeMap type (as in, a map from a type to a value, allowing use of either `==` or `<:`) in julia.h (https://github.com/JuliaLang/julia/compare/jn/methcache?expand=1#diff-67021262df1246d0b5756f73b92eb69fR167)

this may have unintentionally provided a easy fix for #14919 (by deleting a few lines of code, of course :), but that should be a future PR.

vtjnash committed:
introduce an independent method-cache type

this abstracts out the method lookup behavior
of the MethodTable cache
with the goal of making it more re-usable for the tfunc lookup (speed)
and helping address #265 (correctness)

vtjnash committed:
use MethodCache type for tfunc

vtjnash committed:
restore specializations array optimization

vtjnash committed:
make the MethodCache polymorphic

a method cache starts life as jl_nothing, then becomes a method list, then becomes a full cache when it gets big enough

vtjnash committed:
start merging method cache and method lookup implementations more thoroughly

vtjnash committed:
prevent staged functions from messing up the method cache

vtjnash committed:
add missing gc wb for removing conflicting entries from the method table

vtjnash committed:
extend signature lookup map to cover method defs

vtjnash committed:
move type-signature correction to a separate function

this allows us to correct the type signature before passing it to the generated function
or returning it from ml_matches

this also improves the fix for #11840 to cover Unions of the offending types

vtjnash committed:
simplify method cache guard entry decision logic

vtjnash committed:
switch mt->defs over to a full cache

for iterating this, there is now a visitor pattern callback
in both C and Julia

vtjnash committed:
prevent insertion of kinds into the arg1 array

vtjnash committed:
the big rename for TypeMap

Method is temporarily missing now, but will return soon

vtjnash committed:
restore support for guard entries instead of (well, in addition to) guard-signatures

they seemed to be slower, and might be difficult to adapt for #265,
but is worthwhile to have it available to try anyways

vtjnash committed:
the big divide for Method/LambdaInfo

also shares Method more effectively for generated functions
so that instantiating a staged function doesn't make extraneous copies

vtjnash committed:
provide an optimized ml_lookup agent

vtjnash committed:
restore ambiguity warning test

since this did the work in a slightly different order,
a couple existing ambiguity detection flaws were revealed

in ranges, modifying the tvars in the method signature helped to make
the current type system able to resolve this with less ambiguity

in replutil, there was a missing method definition

vtjnash committed:
fix a puzzlingly pre-existing bug that should have been making compile-all unhappy

vtjnash committed:
fix another pre-existing bug, this time in jl_typemap_assoc_exact compuation of args[1-offs]

tkelman commented on base/inference.jl:
```diff
@@ -1630,33 +1583,41 @@ function typeinf_edge(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, nee
+        # toplevel lambda - infer directly
+        frame = InferenceState(linfo, linfo.specTypes, svec(), true)
+        typeinf_loop(frame)
+        @assert frame.inferred # TODO: deal with the better
```
deal with the what?

vtjnash replied:
this. as in the assert that used to be elsewhere, but now is in a place that it can be dealt with

tkelman commented on src/dump.c:
```diff
@@ -785,6 +802,36 @@ static void jl_serialize_value_(ios_t *s, jl_value_t *v)
+        union jl_typemap_t *tf = &m->tfunc;
+        if (tf->unknown && tf->unknown != jl_nothing) {
+            if (m->module == jl_core_module) {
+               tf->unknown = jl_nothing;
```
missing a space

vtjnash commented:
out of curiosity, i tested the change for #14919 (it's only a couple lines to test, although it would take more to make the full change), but it's a bit slower and it runs into #15838 before it can finish compiling. it uncovered a couple of bugs on master though, so i guess that's something?

Keno commented:
There are a number of unrelated commits in here. Can we please get those as a separate pull request? This is enough of a monster to review as is.

vtjnash commented:
only the first two codegen changes are unrelated (they are passing CI, so maybe i should just push those to master now?)

the last two are pre-existing, but conflict significantly with the other changes, so I can't pull them out (although I could commit the equivalent fix to master and then revert them in this PR)

JeffBezanson commented:
Yes, pulling out even 2 commits to master is helpful.

Keno commented:
e6adb2b also looks separate

vtjnash commented:
it could be squashed into ac48b6b, but it's a PR fix, not a bugfix

vtjnash commented:
ok, updated to remove those three commits

Keno commented on src/julia.h:
```diff
@@ -168,68 +187,90 @@ typedef struct _jl_llvm_functions_t {
+    // the most specific for the argument types.
+    union jl_typemap_t invokes;
+
+    int32_t called;  // bit flags: whether each of the first 8 arguments is called
```
Is it still 8 arguments or is it 32 now?

vtjnash replied:
still 8, but now stored in an Int32 type (because that's what's available to jltypes.c during initialization)

Keno commented on base/inference.jl:
```diff
@@ -1630,33 +1583,41 @@ function typeinf_edge(linfo::LambdaInfo, atypes::ANY, sparams::SimpleVector, nee
+        # toplevel lambda - infer directly
+        frame = InferenceState(linfo, linfo.specTypes, svec(), true)
+        typeinf_loop(frame)
+        @assert frame.inferred # TODO: deal with this better
```
Under what condition does this assert fire?

vtjnash replied:
it would require calling a toplevel expr from inside inference, so staged functions that thought it was OK to call `eval` (it ain't)

vtjnash committed:
fix rehashing logic for incremental serializer

plus pack the TypeMapLevel tables when serializing them to maybe save a
bit of space

vtjnash committed:
add some missed gc roots

vtjnash committed:
enable stacktrace-on-assert/abort

also stop hiding the top frame from jl_unw_stepn,
sometimes it's actually useful

vtjnash committed:
Revert "Merge pull request #15860 from JuliaLang/kf/tracing"

This reverts commit b991db1fafa3f1f20b2b95426bb8544ff2a75a7c, reversing
changes made to 9492c5338f2872b2f7204f284f9723bd6b9d2d53.

yuyichao commented:
I'm wondering if the clean up on this branch can help with https://github.com/JuliaLang/julia/issues/15541 ...

vtjnash commented:
i see a small improvement on this branch, but nothing to write home about (95s on v0.4.3, 116s on this branch, 132s on master)

Keno commented:
Why did you revert my branch? Also, this could have used significantly more review and discussion before merging.

JeffBezanson commented on src/julia.h:
```diff
@@ -338,29 +376,41 @@ typedef struct _jl_module_t {
+    jl_array_t *arg1; // Array{union jl_typemap_t}
+    jl_array_t *targ; // Array{union jl_typemap_t}
+    jl_typemap_entry_t *linear; // union jl_typemap_t (but no more levels)
+    jl_value_t *key; // [nullable]
```
What does nullable mean here? C NULL?

vtjnash replied:
yes (taking inspiration from Glib, which uses this notation in its documentation)

JeffBezanson commented on src/julia.h:
```diff
@@ -149,17 +155,30 @@ typedef struct {
+// that can consist of any of the following nodes:
+//   typedef TupleMap Union{TupleMapLevel, TupleMapEntry, Void}
+// it consists of roughly tree-shaped nodes, when possible because `isleaftype(key)`
+// otherwise the leaf entries are stored sorted, linearly
```
I find this comment pretty unclear. Need a more thorough description of the data structure.

JeffBezanson commented on base/reflection.jl:
```diff
@@ -233,28 +233,51 @@ function methods(f::ANY)
     end
 end
 
+function visit(f, mt::MethodTable)
```
I don't see why this is an improvement over being able to iterate over a list of methods. Also `methods` still returns an array sometimes, so now its result is sometimes iterable and sometimes not.

vtjnash replied:
i can implement a variation on `start(mt::MethodTable) = @async visit(produce, mt)` that maintains the current index as a mutable state vector

or an immutable iterator (which would be slightly more expensive, since it needs to recreate a new immutable vector at each step)

do you have a preference for either one?

or i could just define that `methods` always returns an array, which has the benefit of being easier to avoid accidentally mutating as you examine it (e.g. `show(methods(show).cache)`)

JeffBezanson commented:
I think it would help a lot to move the code for typemap to a separate file, since a major point of this change is to factor out the data structure from its clients.

JeffBezanson commented:
I get the sense that using the same structure for method definitions and method/tfunc caches is a big source of complexity here. Passing boolean flags to the table to tell it how to operate is kind of a warning sign to me.

An API cleanup might be in order. For example `jl_typemap_insert` taking 10 arguments is quite scary. Maybe this is an internal function, but then that should be clarified. Overall I'm not sure what the data structure's API is.

There are some un-renamed instances of `TupleMap` remaining. There is also some screwy indentation.

JeffBezanson commented on src/julia_internal.h:
```diff
@@ -151,12 +166,15 @@ jl_value_t *jl_type_match_morespecific(jl_value_t *a, jl_value_t *b);
-                                      jl_lambda_info_t *method, jl_svec_t *tvars,
-                                      int8_t isstaged);
-int jl_is_type(jl_value_t *v);
+void jl_method_table_insert(jl_methtable_t *mt, jl_tupletype_t *type, jl_tupletype_t *simpletype,
```
What does the `simpletype` argument to this mean? It also seems to be more of an implementation detail that doesn't really belong in this entry point.

vtjnash replied:
it fixes the intersections of two bugs on master (#11840 which requires a simple guard entry, and #11355 which prevents the creation of guard entries)

JeffBezanson replied:
Both of those bugs were closed without changing this API. Intuitively, to add a method you need a method table, a method, and possibly a separate type argument. Therefore that's what the arguments should be.

vtjnash replied:
true, they were closed but not fixed. i'll add my test for the remaining case.

JeffBezanson commented on base/inference.jl:
```diff
@@ -716,27 +723,6 @@ const limit_tuple_type_n = function (t::ANY, lim::Int)
     return t
 end
 
-let stagedcache=Dict{Any,Any}()
```
So nice to see this go away.

nalimilan commented:
Is this PR the cause of https://github.com/JuliaLang/julia/issues/15893?

JeffBezanson commented:
Yes, it certainly seems to be.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1489,94 +1451,79 @@ function newvar!(sv::InferenceState, typ)
-    local tfunc_idx = -1
-    local frame
+    local frame = nothing
+    offs = 0
```
What is this for?

vtjnash replied:
compatibility with the existing API inherited from the previous implementation of `MethodTable->cache_targs / cache_args1`. I would like to drop `offs` entirely (I have a commit to do it because it enables another optimization, but it only seemed to save ~0.5% on yichao's dispatch perf test), but in the existing codebase, it is used to decide whether the cache should split on the first or second argument type.

JeffBezanson replied:
Ok, but since inference only ever passes `0` it doesn't seem to need to be part of `jl_tfunc_cache_insert` and `jl_tfunc_cache_lookup`.

JeffBezanson commented on base/interactiveutil.jl:
```diff
@@ -64,7 +64,7 @@ function edit(path::AbstractString, line::Integer=0)
 end
 
-function edit(m::Method)
+function edit(m::TypeMapEntry)
```
I would really like to figure out how to switch this back. Needing to program with `TypeMapEntry` this much seems to go back on the promise of increasing the encapsulation of method caches. We don't want to see packages depending on the fields of this type, which they certainly will if it's returned by functions like `methods` and `methodswith`.

vtjnash replied:
the conundrum i've had is that users probably mostly care about the mapping `sig+tvars -> method` descriptor that this type implements. while the method itself doesn't generally need the `sig` information and instead wants to know specTypes

fwiw, though, this method can trivially be shortened to take a Method since it only cares about file and line number

JeffBezanson replied:
I think the fastest way to fix this is to put a `sig` field back in `Method`, then always return arrays of `Method`s from these functions. Then the cache representation would be completely hidden and I believe we would be backwards-compatible.

Of course, the only small problem then is the redundant `sig` field between Method and TypeMapEntry. To me it looks possible to have TypeMap look inside Method for this when necessary --- especially since TypeMapEntry actually has 2 sig fields, which I suspect could be reduced to 1.

vtjnash replied:
the second field could be removed, but it would require the creation of expensive guard entries (e.g. with tvar environments), so i don't think the performance/memory trade-off there makes sense

> I think the fastest way to fix this is to put a sig field back in Method

also seems acceptable. having a duplicate copy of these fields doesn't seem like the worst thing ever (I don't think it'll be noticeable for performance or memory)

JeffBezanson replied:
Ok, let's do that then. The signature is more or less the most important piece of information about a Method, so it should be easy to see (I believe an example use case just arose in #11242).

JeffBezanson commented:
I do see some nice speedups on some of the tests, such as subarray and linalg/triangular. That's really good, but now I'm getting greedy and want to understand better where the speedup comes from. There are a few candidates:

- Faster tfunc cache lookup
- Faster method dispatch
- Faster `jl_matching_methods`
- Removing `stagedcache`

If we can separate these effects I see potential either for further speedups or simplifying the code or both.

tkelman commented:
CI instability is going to waste more time than faster tests, so if that can't be fixed soon then I think it would be appropriate to revert this for now until it can be done in a stable and properly reviewed way.

And while we're renaming things, what about `gf.c` which is largely getting rewritten here? Always thought that was a cryptic name, could it be expanded a bit?

vtjnash commented:
> I get the sense that using the same structure for method definitions and method/tfunc caches is a big source of complexity here. Passing boolean flags to the table to tell it how to operate is kind of a warning sign to me.

the flags in `jl_typemap_info->unsorted` are there to allow configuration of the cost model, but they should be understood to be an optimization, not a necessity (most of the complexity in this code is about manually unrolling lots of special cases. recreating the algorithm here without optimizations would be much shorter – and slower).

for the other flags, it's tricky since the structure of dispatch is such that it is one data-structure, but there are a number of related questions that we are interested in asking them (intersection, more-specific, typeseq, subtype, and iteration). the bool flags attempt to balance between providing too generic entry points (like `jl_typemap_visitor`) that can't do tree pruning optimizations, and providing separate entry points for each query type (even if the intersection pruning would be the same)

vtjnash commented:
> jl_typemap_insert

this is the entry point for creating a `typemap` entry and inserting it into the map. so it takes all of the arguments for both. we could separate out the creation of the typemap entry (cutting the argument list in half), although i don't see much use right now for a unconnected entry.

vtjnash commented:
> I think it would help a lot to move the code for typemap to a separate file, since a major point of this change is to factor out the data structure from its clients.

agreed. this would involve splitting the file at around line ~1150 and making the the relevant header entries exist. the only part that's made me hesitate is wondering whether gcc is doing any sort of IPO that would be inhibited by the split.

JeffBezanson commented:
> whether gcc is doing any sort of IPO that would be inhibited by the split

That's not the kind of performance I want :P

vtjnash commented:
> If we can separate these effects I see potential either for further speedups or simplifying the code or both.

that depends on which benchmark you want to optimize :P

faster `ml_matches` seemed to be highly effective at driving compile-time tests (and I only did the most basic optimization on it)

faster `jl_method_table_insert` (esp. for batch operations or entire subtrees) should help cut down the time for reloading precompiled modules

faster dispatch? i think we could play more games here to unroll common cases and copy more data inline to avoid a pointer indirection

more compact cache levels? there's a limiting case where the user generates something foolish like `f(::NTuple{100, Int}, ::Val{i})` for `i = 1:32` which will 100 single-entry cache levels in order to build the optimized cache entry table in the last place. it would be more efficient to handle this case by generating one cache level that went from offset 1 to 100, and then a second cache level that contained all of the `Val` entries

andreasnoack commented:
f4f1ee50d8c1e5b3ce406c5bcd3c869d63912052 is causing weird behavior for method definitions in `Calculus.jl` (v0.1.14). Before this commit `differentiate` has 85 methods but after the commit it drops to 14-17 methods depending on which of the commits before a2e4ac4 I'm on.

tkelman commented:
Ah yeah PkgEval would have also been a good idea on a change this big that's supposed to not be user visible.

vtjnash commented:
@andreasnoack fixed by https://github.com/JuliaLang/julia/pull/15904

tkelman commented:
@vtjnash still need to fix `gf.c` not being valid C++
```
c:/projects/julia/src/gf.c(359) : error C2440: 'type cast' : cannot convert from 'jl_value_t *' to 'jl_typemap_t' 
        No constructor could take the source type, or constructor overload resolution was ambiguous 
c:/projects/julia/src/gf.c(413) : error C2440: 'type cast' : cannot convert from 'jl_value_t *' to 'jl_typemap_t' 
        No constructor could take the source type, or constructor overload resolution was ambiguous
```

H-225 commented:
I suspect this PR---specifically commit 35e64c28bfec0bc2c6549491f4724a929cd1bf53 according to Git Blame---has caused https://github.com/JuliaLang/julia/issues/16042, at `REPLCompletions:324`.
